R/pkg/R/column.R

```diff
@@ -169,7 +169,7 @@ setMethod("alias",
 #' @note substr since 1.4.0
 setMethod("substr", signature(x = "Column"),
           function(x, start, stop) {
-            jc <- callJMethod(x@jc, "substr", as.integer(start - 1), as.integer(stop - start + 1))
+            jc <- callJMethod(x@jc, "substr", as.integer(start), as.integer(stop - start + 1))
```

Comment:
I'm a bit concern with changing this. As you can see it's been like this from the very beginning...

Reply:
This API behavior should be considered as wrong and performs inconsistently. Because for starting position 1, we get substring from 1st element, but for position 2, we still get the substring from 1. So we will get the following inconsistent results:

```R
> collect(select(df, substr(df$a, 1, 5)))
  substring(a, 0, 5)
1              abcde
> collect(select(df, substr(df$a, 2, 5)))
  substring(a, 1, 4)
1               abcd
```

For such change, we might need to add a note in the doc as @HyukjinKwon suggested.

Reply:
question:
- is there a way to make the behavior the same before this change for any caller calling substr with common index like 0
- why consider other changes as a follow up and not here? 
https://github.com/apache/spark/pull/20464#issuecomment-362162014

Reply:
> * why consider other changes as a follow up and not here? [#20464 (comment)](https://github.com/apache/spark/pull/20464#issuecomment-362162014)

Just because I think it is another issue regarding 0/negative indices. I can deal it here if you strongly feel it is better.

Reply:
> is there a way to make the behavior the same before this change for any caller calling substr with common index like 0

Should we keep the behavior when calling substr with 0 as start index?

```R
> df <- createDataFrame(list(list(a="abcdef")))
> collect(select(df, substr(df$a, 0, 5)))
  substring(a, -1, 6)
1                   f
> substr("abcdef", 0, 5)
[1] "abcde"
```

I think the previous behavior is pretty unreasonable..